{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5c7a9774-6b64-48c3-8c5b-d16249052972",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "\n",
    "# Bioinformatics Training: Lesson 3\n",
    "\n",
    "## Basic bioinformatics commands: \n",
    "\n",
    "There are tons of built-in commands within Linux OSes that are very useful for bioinformatic analysis. In this section, we will examine a few of the most commonly used commands in bioinformatics. We will adapt content from the Happy Belly Bioinformatics.\n",
    "\n",
    "### `cut`\n",
    "\n",
    "Extract specific columns from tab-delimited files.\n",
    "\n",
    "```bash\n",
    "cut -f1,3 data.tsv\n",
    "# Extracts columns 1 and 3 from a tab-separated file.\n",
    "```\n",
    "\n",
    "### `grep`\n",
    "\n",
    "Search for patterns in a file. Very useful for filtering sequences or logs.\n",
    "\n",
    "```bash\n",
    "grep \"^>\" sequences.fasta\n",
    "# Shows all FASTA headers.\n",
    "```\n",
    "\n",
    "### `paste`\n",
    "\n",
    "Combine files line by line.\n",
    "\n",
    "```bash\n",
    "paste genes.txt lengths.txt\n",
    "# Joins corresponding lines from two files side by side.\n",
    "```\n",
    "\n",
    "### `cat`\n",
    "\n",
    "Concatenate and display file contents.\n",
    "\n",
    "```bash\n",
    "cat sample1.fq sample2.fq > combined.fq\n",
    "# Merges two FASTQ files into one.\n",
    "```\n",
    "\n",
    "### `sed`\n",
    "\n",
    "Stream editor for simple substitutions and line operations.\n",
    "\n",
    "```bash\n",
    "sed 's/chr/Chr/' genome.txt\n",
    "# Replaces 'chr' with 'Chr' in each line.\n",
    "```\n",
    "\n",
    "### `awk`\n",
    "\n",
    "Pattern scanning and processing tool. Useful for complex column-based filtering.\n",
    "\n",
    "```bash\n",
    "awk '$3 > 100' gene_counts.tsv\n",
    "# Shows lines where the third column is greater than 100.\n",
    "```\n",
    "\n",
    "### `tr`\n",
    "\n",
    "Translate or delete characters.\n",
    "\n",
    "```bash\n",
    "tr 'a-z' 'A-Z' < seq.txt\n",
    "# Converts lowercase bases to uppercase.\n",
    "```\n",
    "\n",
    "---\n",
    "\n",
    "## BASH scripting\n",
    "\n",
    "### Variables\n",
    "\n",
    "Variables store and reuse values like filenames or parameters.\n",
    "\n",
    "```bash\n",
    "#!/bin/bash\n",
    "\n",
    "# Define a variable\n",
    "sample=\"sample1\"\n",
    "\n",
    "# Use the variable\n",
    "echo \"The current sample is: $sample\"\n",
    "```\n",
    "\n",
    "### Loops\n",
    "\n",
    "Loops are used to repeat tasks over multiple items like files or values.\n",
    "\n",
    "```bash\n",
    "#!/bin/bash\n",
    "\n",
    "# Loop through all FASTQ files\n",
    "for file in *.fastq; do\n",
    "  echo \"Processing file: $file\"\n",
    "  # Example: run FastQC\n",
    "  fastqc \"$file\"\n",
    "done\n",
    "```\n",
    "\n",
    "### If-Else Statements\n",
    "\n",
    "Conditional logic is helpful to check file presence or handle errors.\n",
    "\n",
    "```bash\n",
    "#!/bin/bash\n",
    "\n",
    "file=\"genome.fasta\"\n",
    "\n",
    "if [ -f \"$file\" ]; then\n",
    "  echo \"$file exists. Proceeding with analysis.\"\n",
    "else\n",
    "  echo \"$file not found. Exiting.\"\n",
    "  exit 1\n",
    "fi\n",
    "```\n",
    "\n",
    "### Scripting\n",
    "\n",
    "A basic pipeline script to organize tasks.\n",
    "\n",
    "```bash\n",
    "#!/bin/bash\n",
    "\n",
    "indir=\"raw_data\"\n",
    "outdir=\"qc_results\"\n",
    "\n",
    "# Create output directory if it doesn't exist\n",
    "mkdir -p \"$outdir\"\n",
    "\n",
    "# Loop through all FASTQ files and run FastQC\n",
    "for file in \"$indir\"/*.fastq; do\n",
    "  if [ -s \"$file\" ]; then\n",
    "    echo \"Running FastQC on $file\"\n",
    "    fastqc \"$file\" -o \"$outdir\"\n",
    "  else\n",
    "    echo \"Warning: $file is empty or missing. Skipping.\"\n",
    "  fi\n",
    "done\n",
    "\n",
    "echo \"Quality check complete for all files.\"\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1a937d0-dceb-47d8-af22-195b72acd583",
   "metadata": {},
   "source": [
    "# Acknowledgement\n",
    "This tutorial is adapted from *Intro to Unix* from Happy Belly Bioinformatics by Michael D. Lee or known as [AstroBioMike in github](https://astrobiomike.github.io/unix/).\n",
    "\n",
    "# Citation\n",
    "[1] Lee, M. (2019). Happy Belly Bioinformatics: an open-source resource dedicated to helping biologists utilize bioinformatics. Journal of Open Source Education, 4(41), 53, https://doi.org/10.21105/jose.00053\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "362aae3d-455f-4a3d-8a52-24a6d19729ab",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": [
     "hide-cell"
    ]
   },
   "outputs": [],
   "source": [
    "%%html\n",
    "<style>\n",
    "# Global Table parameters\n",
    "# table {float:center; width:70%}\n",
    "# tr {float: center; color:black}\n",
    "# th {float: center; color:grey}\n",
    "\n",
    ".term_table table{\n",
    "    float: center;\n",
    "}\n",
    "\n",
    ".term_table table tbody td:nth-child(1){\n",
    "    font-weight: bold;\n",
    "    text-align: center;\n",
    "    width: 200px;\n",
    "    float:center\n",
    "}\n",
    "\n",
    ".term_table table tbody td:nth-child(2){\n",
    "    text-align: center;\n",
    "    width: 300px;\n",
    "    float:center\n",
    "}\n",
    "\n",
    ".term_table table tbody td:nth-child(3){\n",
    "    text-align: center;\n",
    "    width: 300px;\n",
    "    float:center\n",
    "}\n",
    "\n",
    ".term_table table thead tr{\n",
    "    background-color: #b3b3b3;\n",
    "    border: 1px solid black\n",
    "}\n",
    "\n",
    ".term_table table tbody tr:nth-child(even){\n",
    "    background-color:#e5e5e\n",
    "}\n",
    "\n",
    ".image {\n",
    "    margin-left: auto;\n",
    "    margin-right: auto;\n",
    "    width: 50%;\n",
    "}\n",
    "\n",
    ".image2 {\n",
    "    margin-left: auto;\n",
    "    margin-right: auto;\n",
    "    width: 40%;\n",
    "}\n",
    "\n",
    ".image3 {\n",
    "    margin-left: auto;\n",
    "    margin-right: auto;\n",
    "    width: 25%;\n",
    "}\n",
    "\n",
    "</style>\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
