{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Project 2: Strain Heterogeneity Project\n",
    "\n",
    "## Description\n",
    "\n",
    "This project is to analyze the effect of HGT on strain heterogeneity on the binning of the bacterial genomes from metagenome.\n",
    "\n",
    "The hypothesis is that high HGT results causes gene-specific sweeps that results in high strain heterogeneity. Therefore, taxa that exhibits high hgt signals should have high strain heterogeneity and lower quality bins.\n",
    "\n",
    "## <Insert Pipeline>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Strain Analysis\n",
    "\n",
    "Here we will check the heterogeneity per taxa and per bin identified in metagenomes. The metagenomes that will be used are those from 2020-2022 GOWN samples.\n",
    "\n",
    "First we need to identify the amount of heterogeneity per taxa studied. This is measured by *MiDas2/3* and directly identifying strains using *InStrain* Pipeline.\n",
    "\n",
    "\n",
    "### InStrain Pipeline\n",
    "\n",
    "Here let us run the InStrain Pipeline. First we need to build a genome database to which the metagenomes will be compared too.\n",
    "\n",
    "<ins>**Dereplicate**</ins> the intraspecific species genomes with mash at ANI: 98% (I am looking at 99.2% which equivalent to 1.2 bp difference in a 150 long read). A 98% similarity is equivalent to read similarity of 3bp difference in a 150 bp long read.\n",
    "\n",
    "Dereplication was done in three modes:\n",
    "1. Bin Only\n",
    "2. GTDB Only\n",
    "3. GTDB + Bin\n",
    "\n",
    "To dereplicate bins only, the following code was run from `/home/gamaliellysandergl.c/p2_workstation/drep`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# drep{1}.slurm\n",
    "#!/bin/bash\n",
    "# ---------------------------------------------------------------------\n",
    "# An example PBS script for running a job on a compute cluster\n",
    "# ---------------------------------------------------------------------\n",
    "#SBATCH --job-name=dRepBin98\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --cpus-per-task=16\n",
    "#SBATCH --mem=240gb\n",
    "#SBATCH --time=7-00:00:00\n",
    "\n",
    "# ---------------------------------------------------------------------\n",
    "echo \"Starting run at: `date`\"\n",
    "# ---------------------------------------------------------------------\n",
    "\n",
    "source ~/software/miniforge3/etc/profile.d/conda.sh \n",
    "conda activate instrain\n",
    "\n",
    "workdir='/home/gamaliellysandergl.c/p2_workstation'\n",
    "\n",
    "### 01 - Preparing for Drep ###\n",
    "# Import the reference genome from the GTDB downloaded database.\n",
    "# This version uses the GTDB R214 release\n",
    "# python instrain_ref_fetch.py instrain_taxa_list.txt\n",
    "# bash instrain_bin_fetch.sh instrain_bin_list.txt\n",
    "# echo \"inStrain GTDB reference database concatenate at `date`\"\n",
    "\n",
    "### 02 - Drep Run ###\n",
    "dRep dereplicate BinGenomeSet -g /home/gamaliellysandergl.c/p2_workstation/bins/*.fa \\\n",
    "--completeness 50 \\\n",
    "--checkM_method taxonomy_wf \\\n",
    "--S_algorithm fastANI --multiround_primary_clustering \\\n",
    "--clusterAlg average \\\n",
    "-ms 10000 \\\n",
    "-pa 0.9 -sa 0.98 \\\n",
    "-nc 0.30 -cm larger -p 16 \\\n",
    "-d\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The results can be find in: `/home/gamaliellysandergl.c/p2_workstation/drep/BinGenomeSet`\n",
    "\n",
    "A different version which combines the GTDB Genome Set and Bin Genome Set into one is made by `drep3.slurm`. The output folder is named as `MergedGenomeSet`. The code is here:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#drep3.slurm\n",
    "\n",
    "## <insert SBATCH parameters here>\n",
    "\n",
    "# --------------------------------------------------------------------\n",
    "echo \"Starting run at: `date`\"\n",
    "# --------------------------------------------------------------------\n",
    "\n",
    "source ~/software/miniforge3/etc/profile.d/conda.sh \n",
    "conda activate instrain\n",
    "\n",
    "workdir='/work/ebg_lab/referenceDatabases/dRep'\n",
    "scratch=/scratch/$SLURM_JOBID\n",
    "\n",
    "\n",
    "### 01 - Preparing for Drep ###\n",
    "# Import the reference genome from the GTDB downloaded database.\n",
    "# This version uses the GTDB R214 release\n",
    "# python instrain_ref_fetch.py instrain_taxa_list.txt\n",
    "# bash instrain_bin_fetch.sh instrain_bin_list.txt\n",
    "# echo \"inStrain GTDB reference database concatenate at `date`\"\n",
    "\n",
    "# Decompress *.gz files in scratch\n",
    "while IFS= read -r filepath; do\n",
    "filename=${filepath##*/} # same as $(basename \"$filepath\")\n",
    "prefix=$(echo \"$filename\" | rev | cut -d'.' -f3- | rev)\n",
    "newname=\"${prefix}.fa\"\n",
    "\n",
    "cp ${filepath} ${scratch}/${filename}\n",
    "gzip -d \"${scratch}/${filename}\"\n",
    "\n",
    "mv ${scratch}/${filename%.*} ${scratch}/${newname}\n",
    "\n",
    "done < ref_genome_path.txt\n",
    "\n",
    "echo \"`ls ${scratch}`\"\n",
    "\n",
    "# Move Bins to scratch\n",
    "cp /home/gamaliellysandergl.c/p2_workstation/bins/*.fa ${scratch}\n",
    "\n",
    "\n",
    "### 02 - Drep Run ###\n",
    "dRep dereplicate ${scratch}/MergedGenomeSet -g ${scratch}/*.fa \\\n",
    "--completeness 50 \\\n",
    "--checkM_method taxonomy_wf \\\n",
    "--S_algorithm fastANI --multiround_primary_clustering \\\n",
    "--clusterAlg average \\\n",
    "-ms 10000 \\\n",
    "-pa 0.9 -sa 0.98 \\\n",
    "-nc 0.30 -cm larger -p 16 \\\n",
    "-d\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The outputs for unique genomes for all Genome Sets can be found in `drep/*GenomeSet/dereplicated_genomes`. The content of this folder are same format and name as the input. \n",
    "\n",
    "<ins>**Concatenation**</ins> of these genomes into a single fasta is needed to make the reference for inStrain. Unfortunately, Bin Genomes sometimes have the same headers between different fasta files so it can create errors during inStrain. Therefore, prior to concatenation, we will rename the headers by prefixing the headers with the genome/fasta filename (i.g. *>O10.maxbin.2.fa--K141_123124*). The code for concatenation is down below:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#drep_concatenate.slurm\n",
    "\n",
    "#!/bin/bash\n",
    "# ---------------------------------------------------------------------\n",
    "# An example PBS script for running a job on a compute cluster\n",
    "# ---------------------------------------------------------------------\n",
    "#SBATCH --job-name=dRepcat\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --cpus-per-task=4\n",
    "#SBATCH --mem=24gb\n",
    "#SBATCH --time=7-00:00:00\n",
    "# ---------------------------------------------------------------------\n",
    "echo \"Starting run at: `date`\"\n",
    "# ---------------------------------------------------------------------\n",
    "\n",
    "source ~/software/miniforge3/etc/profile.d/conda.sh \n",
    "conda activate instrain\n",
    "\n",
    "for i in `ls -d *GenomeSet`\n",
    "do\n",
    "workdir=${i}/dereplicated_genomes\n",
    "\n",
    "    for j in `ls ${workdir}/*.fa`\n",
    "    do\n",
    "    filename=${j##*/}\n",
    "    sed -i \"s/^>/>${filename}--/g\" ${j}\n",
    "    done #j\n",
    "\n",
    "cat ${workdir}/*.fa > ${i}/${i}_concatenated.fa\n",
    "\n",
    "done #i"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The concatenated genome will be a single file inside the Genome Set folders. \n",
    "\n",
    "Now that the reference genomes has been dereplicated and concatenated, we will process the reference fasta file through <ins>**parse_stb**</ins>, building <ins>**bowtie index**</ins>, creating a bam file using <ins>**bowtie**</ins> it self, and predicting genes using <ins>**prodigal**</ins> .\n",
    "\n",
    "<ins>**Parsing**</ins> the input genomes into a scaffold-to-bin (stb) file allows back tracking of the contigs to its origin and needed by inStrain. The `drep_parse_stb.slurm` uses the `parse_stb.py` that comes along drep installation to create the stb file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "%%bash\n",
    "#!/bin/bash\n",
    "# ---------------------------------------------------------------------\n",
    "# An example PBS script for running a job on a compute cluster\n",
    "# ---------------------------------------------------------------------\n",
    "#SBATCH --job-name=dRepstb\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --cpus-per-task=4\n",
    "#SBATCH --mem=24gb\n",
    "#SBATCH --time=7-00:00:00\n",
    "\n",
    "# ---------------------------------------------------------------------\n",
    "echo \"Starting run at: `date`\"\n",
    "# ---------------------------------------------------------------------\n",
    "\n",
    "source ~/software/miniforge3/etc/profile.d/conda.sh \n",
    "conda activate instrain\n",
    "\n",
    "for i in `ls -d *GenomeSet`\n",
    "do\n",
    "echo $i\n",
    "parse_stb.py --reverse -f ${i}/dereplicated_genomes/*.fa \\\n",
    "-o ${i}_parse.stb \n",
    "\n",
    "mv ${i}_parse.stb ${i}/\n",
    "\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The stb files are moved into individual Genome Set folders with the filename is the Genome Set name: `*GenomeSet_parse.stb`.\n",
    "\n",
    "<ins>**Bowtie index**</ins> was build using bowtie2-build default options. Unlike the prior `slurm` files, bowtie indexing were done inside the `instrain2` folder which is adjacent to `drep` folder. The `instrain2` folder contains the `bowtie` folder and inside `bowtie` is `bowtie-build` which contains the indices. The script is here:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#instrain_1_bowtie.slurm\n",
    "\n",
    "#!/bin/bash\n",
    "# ---------------------------------------------------------------------\n",
    "# An example PBS script for running a job on a compute cluster\n",
    "# ---------------------------------------------------------------------\n",
    "#SBATCH --job-name=instBT2build\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --cpus-per-task=16\n",
    "#SBATCH --mem=64gb\n",
    "#SBATCH --time=7-00:00:00\n",
    "\n",
    "# ---------------------------------------------------------------------\n",
    "echo \"Starting run at: `date`\"\n",
    "# ---------------------------------------------------------------------\n",
    "\n",
    "source ~/software/miniforge3/etc/profile.d/conda.sh \n",
    "conda activate bowtie\n",
    "\n",
    "workdir=${PWD}/bowtie/bowtie2-build\n",
    "mkdir -p ${workdir}\n",
    "cd ${workdir}\n",
    "\n",
    "for i in `ls -d /home/gamaliellysandergl.c/p2_workstation/drep/*Set`\n",
    "do\n",
    "filename=${i##*/}\n",
    "echo ${filename}\n",
    "bowtie2-build ${i}/${filename}_concatenated.fa ${filename}_concatenated.fa \n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After the indices are built, the main <ins>**bowtie**</ins> run will be run. The `*.sam` files are located inside the `bowtie` folder with the name `<Reference GenomeSet>_v_<Metagenome Reads>.sam`. The code for it is:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#instrain_2_bowtie.slurm\n",
    "\n",
    "#!/bin/bash\n",
    "# ---------------------------------------------------------------------\n",
    "# An example PBS script for running a job on a compute cluster\n",
    "# ---------------------------------------------------------------------\n",
    "#SBATCH --job-name=inStrBT2\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --cpus-per-task=20\n",
    "#SBATCH --mem=80gb\n",
    "#SBATCH --time=7-00:00:00\n",
    "\n",
    "# ---------------------------------------------------------------------\n",
    "echo \"Starting run at: `date`\"\n",
    "# ---------------------------------------------------------------------\n",
    "\n",
    "source ~/software/miniforge3/etc/profile.d/conda.sh \n",
    "conda activate bowtie\n",
    "\n",
    "# PWD run inside instrain2\n",
    "\n",
    "workdir=${PWD}/bowtie/bowtie2-build\n",
    "outputdir=${PWD}/bowtie\n",
    "mkdir -p ${workdir}\n",
    "\n",
    "\n",
    "# Fetch the metagenomes\n",
    "\n",
    "for j in `ls -d ~/p2_workstation/instrain2/metagenomes/AOB*/`\n",
    "do\n",
    "metagenome=$(echo $j | rev | cut -d/ -f2 | rev)\n",
    "R1=\"`ls ${j}/*qc.R1.fastq`\"\n",
    "R2=\"`ls ${j}/*qc.R2.fastq`\"\n",
    "\n",
    "# Fetch the references\n",
    "for i in `ls -d /home/gamaliellysandergl.c/p2_workstation/drep/*Set`\n",
    "do\n",
    "filename=${i##*/} #filename is either BinGenomeSet, GTDBGenomeSet, or MergedGenomeSet\n",
    "output_file=\"${filename}_v_${metagenome}\"\n",
    "echo \"# Running: ${output_file}\"\n",
    "\n",
    "bowtie2 -p $SLURM_CPUS_PER_TASK \\\n",
    "-x ${workdir}/${filename}_concatenated.fa \\\n",
    "-1 ${R1} -2 ${R2} > \\\n",
    "${outputdir}/${output_file}.sam\n",
    "done\n",
    "\n",
    "done #j\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "On the side, <ins>**Prodigal**</ins> can be run simultaneously with the *bowtie* slurm files without dependencies. The prodigal files will be in the `drep/*GenomeSet/prodigal`  folder outside of `instrain` folder. The it should contain `*genes.fna` and `*genes.faa` files. The script:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#instrain_3_prodigal.slurm\n",
    "\n",
    "#!/bin/bash\n",
    "# ---------------------------------------------------------------------\n",
    "# An example PBS script for running a job on a compute cluster\n",
    "# ---------------------------------------------------------------------\n",
    "#SBATCH --job-name=instrprodigal\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --cpus-per-task=8\n",
    "#SBATCH --mem=24gb\n",
    "#SBATCH --time=7-00:00:00\n",
    "\n",
    "# ---------------------------------------------------------------------\n",
    "echo \"Starting run at: `date`\"\n",
    "# ---------------------------------------------------------------------\n",
    "\n",
    "source ~/software/miniforge3/etc/profile.d/conda.sh \n",
    "conda activate instrain\n",
    "\n",
    "for i in `ls -d ~/p2_workstation/drep/*Set`\n",
    "do\n",
    "GenomeSet=${i##*/}\n",
    "echo ${GenomeSet}\n",
    "\n",
    "mkdir -p ${i}/prodigal\n",
    "\n",
    "#echo \" \\\n",
    "prodigal -i ${i}/${GenomeSet}_concatenated.fa \\\n",
    "-d ${i}/prodigal/${GenomeSet}_concatenated.fa.genes.fna \\\n",
    "-a ${i}/prodigal/${GenomeSet}_concatenated.fa.genes.faa\n",
    "#\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With all the above preparation done, we now run the `inStrain profile` command to get the initial IS profile output of InStrain. The command we use is this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#instrain_4_profile.slurm\n",
    "\n",
    "#!/bin/bash\n",
    "# ---------------------------------------------------------------------\n",
    "# An example PBS script for running a job on a compute cluster\n",
    "# ---------------------------------------------------------------------\n",
    "#SBATCH --job-name=inStrProfile\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --cpus-per-task=32\n",
    "#SBATCH --mem=128gb\n",
    "#SBATCH --time=7-00:00:00\n",
    "\n",
    "# ---------------------------------------------------------------------\n",
    "echo \"Starting run at: `date`\"\n",
    "# ---------------------------------------------------------------------\n",
    "\n",
    "source ~/software/miniforge3/etc/profile.d/conda.sh \n",
    "conda activate instrain\n",
    "\n",
    "# PWD run inside instrain2\n",
    "samdir=${PWD}/bowtie\n",
    "profile_dir=${PWD}/profile\n",
    "mkdir -p ${profile_dir}\n",
    "\n",
    "# Fetch the metagenomes\n",
    "\n",
    "for j in `ls -d ~/p2_workstation/instrain2/metagenomes/AOB*/`\n",
    "do\n",
    "metagenome=$(echo $j | rev | cut -d/ -f2 | rev)\n",
    "\n",
    "# Fetch the references\n",
    "for i in `ls -d /home/gamaliellysandergl.c/p2_workstation/drep/*Set`\n",
    "do\n",
    "GenomeSet=${i##*/} #GenomeSet is either BinGenomeSet, GTDBGenomeSet, or MergedGenomeSet\n",
    "bam_file=\"${GenomeSet}_v_${metagenome}\"\n",
    "fasta_file=\"`ls ${i}/*concatenated.fa`\"\n",
    "genes=\"`ls ${i}/prodigal/*genes.fna`\"\n",
    "parse_stb=\"`ls ${i}/*.stb`\"\n",
    "echo \"# Running: ${bam_file}\"\n",
    "\n",
    "#echo \"\\\n",
    "inStrain profile \\\n",
    "${samdir}/${bam_file}.sam \\\n",
    "${fasta_file} \\\n",
    "-p $SLURM_CPUS_PER_TASK \\\n",
    "-o ${profile_dir}/${bam_file}.IS \\\n",
    "-g ${genes} \\\n",
    "-s ${parse_stb}\n",
    "#\"\n",
    "\n",
    "done\n",
    "\n",
    "done #j"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each sam files was converted to bam files. The output profile is created per each bam file. Thus, the amount of profile created depends on `No. of GenomeSet * No. of Metagenomes`. The IS profiles output are one per combination mentioned. It is a folder and more details on its output can be found on this [inStrain documentation](https://instrain.readthedocs.io/en/latest/).\n",
    "\n",
    "#### Analysis of inStrain profile"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "openai-api",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
